examples/exex/rollup/src/db.rs

```diff
@@ -0,0 +1,482 @@
+        {
+            tx.execute(
+                "INSERT INTO account_revert (block_number, address, data) VALUES (?, ?, ?) ON CONFLICT(block_number, address) DO UPDATE SET data = excluded.data",
+                (block.header.number.to_string(), address.to_string(), serde_json::to_string(account)?),
+            )?;
+        }
+
+        for PlainStorageChangeset { address, wipe_storage, storage } in changeset.storage {
+            if wipe_storage {
+                tx.execute("DELETE FROM storage WHERE address = ?", (address.to_string(),))?;
+            }
+
+            for (key, data) in storage {
+                tx.execute(
+                    "INSERT INTO storage (address, key, data) VALUES (?, ?, ?) ON CONFLICT(address, key) DO UPDATE SET data = excluded.data",
+                    (address.to_string(), B256::from(key).to_string(), data.to_string()),
+                )?;
+            }
+        }
+
+        if reverts.storage.len() > 1 {
+            eyre::bail!("too many blocks in storage reverts");
+        }
+        for PlainStorageRevert { address, wiped, storage_revert } in
+            reverts.storage.into_iter().next().ok_or(eyre::eyre!("no storage reverts"))?
+        {
+            let storage = storage_revert
+                .into_iter()
+                .map(|(k, v)| (B256::new(k.to_be_bytes()), v))
+                .collect::<Vec<_>>();
+            let wiped_storage = if wiped { get_storages(&tx, address)? } else { Vec::new() };
+            for (key, data) in StorageRevertsIter::new(storage, wiped_storage) {
+                tx.execute(
+                    "INSERT INTO storage_revert (block_number, address, key, data) VALUES (?, ?, ?, ?) ON CONFLICT(block_number, address, key) DO UPDATE SET data = excluded.data",
+                    (block.header.number.to_string(), address.to_string(), key.to_string(), data.to_string()),
+                )?;
+            }
+        }
+
+        for (hash, bytecode) in changeset.contracts {
+            tx.execute(
+                "INSERT INTO bytecode (hash, data) VALUES (?, ?) ON CONFLICT(hash) DO NOTHING",
+                (hash.to_string(), bytecode.bytes().to_string()),
+            )?;
+        }
+
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    /// Revert block from the database.
+    ///
+    /// The code is adapted from <https://github.com/paradigmxyz/reth/blob/a8cd1f71a03c773c24659fc28bfed2ba5f2bd97b/crates/storage/provider/src/providers/database/provider.rs#L365-L540>
+    pub fn revert_block(&mut self, number: u64) -> eyre::Result<()> {
+        let mut connection = self.connection();
+        let tx = connection.transaction()?;
+
+        tx.execute("DELETE FROM block WHERE number = ?", (number.to_string(),))?;
+
+        let mut state = BundleStateInit::new();
+        let mut reverts = RevertsInit::new();
+
+        let account_reverts = tx
+            .prepare("SELECT address, data FROM account_revert WHERE block_number = ?")?
```

Comment:
agreed, I made the function signature and the comment explicit about being able to revert only the tip block, so that on a re-org we revert from the tip block by block.

Reply:
sg